### ISSUE TITLE
Parallelize Test Runs

### ISSUE THREAD

**Issue by rtfeldman** [rtfeldman/node-test-runner#45]
## Design

1. Have the CLI compile `elmTestWorker.js` and then take on the role of supervisor.
2. Supervisor spawns up to 4 `node` processes (~10mb overhead each), each running `elmTestWorker.js`
3. On startup, each worker distributes seeds and sorts the pending tests into a `pending` data structure. (Supervisor sends each worker a seed, either randomly generated or passed through from `--seed` CLI flag, so that they all independently arrive at identical `pending` data structures.)
3. The supervisor sends each one a test index to run, e.g. "worker 1: run test 1, worker 2: run test 2", etc.
4. Each worker finds the appropriate index in the `pending` data structure and runs it.
5. When a worker finishes, it reports the result back to the supervisor, and then waits for further instructions.
6. When the supervisor hears back from a worker, it adds that worker's result to a `results` map.
7. When the `results` map has contiguous unreported tests (e.g. tests 5-7 have all come back), it reports them to the end user (that is, prints them on the screen). Uncontiguous ones (e.g. 7 has come in, but still waiting on 5 and 6) means to hold off on reporting. This will make it so the output order is the same every time, regardless of what's going on in parallel.
8. Once a worker's response has been processed, the supervisor increments the current test index and sends it to that worker.
9. If a worker gets a text index that's out of bounds, that means there are no more tests to run, so the worker process stops waiting for messages and exits.
10. Once all workers have exited, the supervisor is done and can print a summary.

Since only the workers need to interact with Elm code, the CLI/supervisor can be written in any language.

### Worker
* Receive configuration options (including the `.js` file to run) in `argv`
* Listen on `stdin` for integers corresponding to test indices
* Echo JSON test results after running each test
* When the received index is out of bounds, we're done; exit 0
* If any errors happen, exit 1

### Supervisor
* Accept a `--path-to-node` CLI flag in case you don't want to use the version of `node` that's on your `PATH` for the workers. This would also enable testing out custom worker variants; instead of pointing it at a vanilla `node` binary, point it at something that does something fancier with the extremely basic communication protocol. (Default value: the `node` on your `PATH`)
* Detect # of cores available; spawn 1 `node` process per core
* Follow the steps as described above

## Other Thoughts

https://github.com/rtfeldman/elm-web-workers doesn't seem necessary on Node, although we might want to use it to parallelize the tests in the browser.

**Comment by gampleman** [comment]
I wonder if something that should be considered for the design is how this could work for CI environments, where you would want to distribute test runs across multiple machines/containers.

**Comment by zkessin** [comment]
Data point, at least on Circle CI it would be nice to have an option to have it run on 1 container (as that is free) or across several depending on the size of the suite and how much people feel like paying
